feat(Idempotency): add feature for manipulating idempotent responses (#4037)
awslabs/aws-lambda-powertools-python: docs/utilities/idempotency.md

    participant Persistence Layer
    alt initial request
        Client->>Lambda: Invoke (event)
        Lambda->>Persistence Layer: Get or set idempotency_key=hash(payload)
        activate Persistence Layer
        Note over Lambda,Persistence Layer: Set record status to INPROGRESS. <br> Prevents concurrent invocations <br> with the same payload
        Lambda-->>Lambda: Call your function
        Lambda->>Persistence Layer: Update record with result
        deactivate Persistence Layer
        Persistence Layer-->>Persistence Layer: Update record
        Note over Lambda,Persistence Layer: Set record status to COMPLETE. <br> New invocations with the same payload <br> now return the same result
        Lambda-->>Client: Response sent to client
    else retried request
        Client->>Lambda: Invoke (event)
        Lambda->>Persistence Layer: Get or set idempotency_key=hash(payload)
        activate Persistence Layer
        Persistence Layer-->>Response hook: Already exists in persistence layer.
        deactivate Persistence Layer
        Note over Response hook,Persistence Layer: Record status is COMPLETE and not expired
        Response hook->>Lambda: Response hook invoked

>         Lambda-->>Client: Same response sent to client

Reviewer: ```suggestion
        Lambda-->>Client: Manipulated idempotent response sent to client
```

Otherwise it makes me thing that the Response Hook can't change anything

    end
```

> <i>Idempotent successful request with response hook</i>

Reviewer: ```suggestion
<i>Successful idempotent request with a response hook</i>
```

| **expires_after_seconds**       | 3600    | The number of seconds to wait before a record is expired                                                                                                                                                                            |
| **use_local_cache**             | `False` | Whether to locally cache idempotency results                                                                                                                                                                                        |
| **local_cache_max_items**       | 256     | Max number of items to store in local cache                                                                                                                                                                                         |
| **hash_function**               | `md5`   | Function to use for calculating hashes, as provided by [hashlib](https://docs.python.org/3/library/hashlib.html){target="_blank" rel="nofollow"} in the standard library.                                                           |

> | **response_hook**               | `None`  | Function to use for processing the stored Idempotent response.  This function hook is called when an already returned response is found.  See [Modifying The Idempotent Response](idempotency.md#modifying-the-idempotent-repsonse) |

Reviewer: ```suggestion
| **response_hook**               | `None`  | Function to use for processing the stored Idempotent response.  This function hook is called when an existing idempontent response is returned.  See [Modifying The Idempotent Response](idempotency.md#modifying-the-idempotent-repsonse) |
```

    For example, the `_put_record` method needs to raise an exception if a non-expired record already exists in the data store with a matching key.

> ### Modifying the Idempotent Repsonse

Reviewer 2: Not everyone will want to modify the response. Perhaps `Accessing the Idempotent Response` would be more appropriate?

Reviewer: What about `Manipulating the Idempotent Response`?

    For example, the `_put_record` method needs to raise an exception if a non-expired record already exists in the data store with a matching key.

### Manipulating the Idempotent Response

> You can set up a `response_hook` in the `IdempotentConfig` class to access the returned data when an operation is idempotent. The hook function will be called with the current deserialized response object and the Idempotency record.

Reviewer: ```suggestion
You can set up a `response_hook` in the `IdempotentConfig` class to manipulate the returned data when an operation is idempotent. The hook function will be called with the current deserialized response object and the Idempotency record.
```

=== "Using an Idempotent Response Hook"

    ```python hl_lines="18 20 23 32"
    --8<-- "examples/idempotency/src/working_with_response_hook.py"
    ```

=== "Sample event"

    ```json
    --8<-- "examples/idempotency/src/working_with_response_hook_payload.json"
    ```

???+ info "Info: Using custom de-serialization?"

    The response_hook is called after the custom de-serialization so the payload you process will be the de-serialized version.

#### Being a good citizen

> Using Response hooks can add subtle improvements to manipulating returned data from idempotent operations, but also add significant complexity if you're not careful.
> 
> Keep the following in mind when authoring hooks for Idempotency utility:
> 
> 1. **Response hook works exclusively when operations are idempotent.** Carefully consider the logic within the `Response hook` and prevent any attempt to access the key from relying exclusively on idempotent operations.
> 
> 2. **Catch your own exceptions.** Catch and handle known exceptions to your logic.
> 
> 3. **Watch out when you are decorating the Lambda Handler and using the Response hook.** If you don't catch and handle exceptions in your `Response hook`, your function might not run properly.

Reviewer: A suggestion to make this more clear:

1- response hooks only work on idempotent responses
2- catch exceptions
3- keep it simple

Is this better?

```suggestion
When using response hooks to manipulate returned data from idempotent operations, it's important to follow best practices to avoid introducing complexity or issues. Keep these guidelines in mind:

1. **Response hook works exclusively when operations are idempotent.** The hook will not be called when an operation is not idempotent, or when the idempotent logic fails.

2. **Catch and Handle Exceptions** Your response hook code should catch and handle any exceptions that may arise from your logic. Unhandled exceptions will cause the Lambda function to fail unexpectedly.

3. **Keep Hook Logic Simple** Response hooks should contain minimal, straightforward logic for manipulating response data. Avoid complex conditional branching and aim for hooks that are easy to reason about.
```

Reviewer 2: This is much better than the one I wrote! Thank you for this very useful review and suggestion @rubenfonseca
